Workflows/Rails upgrades.md

```diff
@@ -0,0 +1,286 @@
+- sync with the devops team to provide a Ruby installation estimate if applicable
+
+Projects in bad health will require extra care in the estimation phase. The following factors may severly (e.g. 3x) impact the estimate
+
+- low test coverage
+- low internal application functionality knowledge
+- being more than 1 major Rails or Ruby version behind
+- incomplete or proprietary deployment pipeline
+- poor codebase health
+- lack of functional specification
+- forked dependencies
+- unmaintained dependencies (e.g. paperclip)
+- the app is not bootable locally
+- there is no staging environment
+
+**Please perform a sanity check with a TL/LE for all major Rails upgrades.**
+
+## Upgrade steps
+The upgrade process can be split into multiple smaller phases. Splitting the work, reviewing and deploying it individually enables us to:
+
+- introduce smaller code changes,
+- observe and review a single change at a time,
+- perform smaller, more targeted smoke tests of the application,
+- more easily plan multiple smaller amounts of work (usually a a day or so per phase) as opposed to a big (weeks) chunk of work,
+- estimate the work with a higher degree of confidence.
+
+The phases noted below should each contain their own task, acceptance criteria and verification steps (e.g. partial or full smoke test). They should be deployed to the production environment **individually** and there should be some grace period between them when required.
+
+All rules have exceptions so please check in with a fellow TL/LE if you won't be following these guidelines due to project specific constraints. Bundling the upgrade steps together does introduce extra risk so we should be explicit and communicate the tradeoffs to stakeholders.
+
+### 1. Upgrading gems for Rails X support
+Rails has a strong gravitational pull towards the general gem ecosystem, so it is very likely that most Rails upgrades will require some auxiliary dependencies to be upgraded as well. Compiling a list of gems that need to be upgraded to support the next Rails version will allow you to split the upgrade into multiple parts and deploy smaller, safer changes.
+
+Gems declare their dependency version constraints in the `gemspec` file, but that is not a solid guarantee that a dependency supports the next version of Rails.
+
+#### Explicit dependency Rails support
+All project dependencies and their dependency requirements are listed in the `Gemfile.lock` file. Some gems constrict their dependency requirements quite conservatively while others keep an open mind. Relaxing these dependencies (usually by upgrading them) is the first step towards enabling you to upgrade to the next Rails version.
+
+In the following case we present a situation where a project is using Rails 6.1, but is blocked by one of the dependencies (`delayed_job`) to versions betwen 3.0 and 6.0.x.
```

Comment:
Hmm, maybe this wording is a bit weird. Your app is not using Rails 6.1, but you're rather attempting to upgrade your app to Rails 6.1?